File: axis-ref/backend copy/axis/decision/evaluator.py
"""Decision evaluator.

Pure function: `evaluate(action, region) -> Outcome`. The TS mirror in
`frontend/src/decision/evaluator.ts` MUST produce the same numbers; both
read the same Action blob (shipped via `state.json`) and the same
RegionIntel blob (shipped via `intel.json`).

Formula and constants are documented in `docs/decision-engine.md`.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Iterable, Literal

from axis.decision.actions import Action
from axis.intel.events import EventCategory
from axis.intel.morale import RegionIntel

SEVERITY_DIVISOR = 12.0
CONTRIBUTION_DIVISOR = 8.0
P_FLOOR = 0.05
P_CEIL = 0.95
SIGNIFICANT_DELTA = 0.015
# Group-level scalars applied at evaluation time. Intel-side contributions
# (morale, trend, severity, category drivers) are dampened; political-side
# contributions (pressure, credibility) are amplified, so political context
# outweighs morale by roughly 3:1 in per-action max swing. The TS mirror in
# `frontend/src/decision/evaluator.ts` MUST keep the same values.
MORALE_SCALE = 1.0 / 3.0
POLITICAL_SCALE = 3.0


BreakdownKind = Literal["base", "modifier", "category"]


@dataclass(frozen=True, slots=True)
class BreakdownItem:
    label: str
    kind: BreakdownKind
    delta: float  # contribution to probability, signed
    # Stable identifier the explain endpoint uses to attach a payload to a
    # specific row. Examples: "base", "morale", "trend", "severity",
    # "cat:protest", "pressure", "credibility". Optional for legacy callers.
    key: str = ""

    def to_dict(self) -> dict[str, object]:
        d: dict[str, object] = {
            "label": self.label,
            "kind": self.kind,
            "delta": round(self.delta, 4),
        }
        if self.key:
            d["key"] = self.key
        return d


@dataclass(frozen=True, slots=True)
class Outcome:
    action_id: str
    region_id: str
    probability: float
    breakdown: tuple[BreakdownItem, ...]
    explanation: str

    def to_dict(self) -> dict[str, object]:
        return {
            "action_id": self.action_id,
            "region_id": self.region_id,
            "probability": round(self.probability, 4),
            "breakdown": [b.to_dict() for b in self.breakdown],
            "explanation": self.explanation,
        }


def _clamp(x: float, lo: float, hi: float) -> float:
    if x < lo:
        return lo
    if x > hi:
        return hi
    return x


def _trend_signed(label: str) -> float:
    if label == "rising":
        return 1.0
    if label == "declining":
        return -1.0
    return 0.0


@dataclass(frozen=True, slots=True)
class PoliticalContext:
    """Political layer scalars the evaluator factors into probability.

    All fields are optional; absent values are treated as zero so legacy
    callers and intel-only contexts still work.
    """

    issuer_pressure: float | None = None  # 0..1 issuer faction pressure
    issuer_deadline_turns_remaining: int | None = None  # informational only
    bilateral_credibility_immediate: float | None = None  # -1..+1
    bilateral_credibility_resolve: float | None = None  # -1..+1
    issuer_faction_id: str | None = None
    target_faction_id: str | None = None


def evaluate(
    action: Action,
    region: RegionIntel,
    political: PoliticalContext | None = None,
) -> Outcome:
    morale_norm = (region.morale_score - 50.0) / 50.0
    p_morale = morale_norm * action.morale_weight * MORALE_SCALE

    trend_signed = _trend_signed(region.morale_trend)
    p_trend = trend_signed * action.trend_weight * MORALE_SCALE

    severity_sum = sum(d.contribution for d in region.drivers)
    severity_norm = _clamp(severity_sum / SEVERITY_DIVISOR, -1.0, 1.0)
    p_severity = severity_norm * action.severity_weight * MORALE_SCALE

    category_items: list[BreakdownItem] = []
    for d in region.drivers:
        sensitivity = action.category_sensitivities.get(d.category, 0.0)
        if sensitivity == 0.0:
            continue
        # Intensity is loudness (always non-negative); the catalog's signed
        # sensitivity decides whether this category helps or hurts the action.
        intensity = _clamp(abs(d.contribution) / CONTRIBUTION_DIVISOR, 0.0, 1.0)
        delta = sensitivity * intensity * MORALE_SCALE
        if abs(delta) < SIGNIFICANT_DELTA:
            continue
        category_items.append(
            BreakdownItem(
                label=_category_label(d.category, d.contribution),
                kind="category",
                delta=delta,
                key=f"cat:{d.category.value}",
            )
        )

    p_categories = sum(item.delta for item in category_items)

    p_pressure, pressure_label = _pressure_delta(action, political)
    p_credibility, credibility_label = _credibility_delta(action, political)
    p_pressure *= POLITICAL_SCALE
    p_credibility *= POLITICAL_SCALE

    raw = (
        action.base_rate
        + p_morale
        + p_trend
        + p_severity
        + p_categories
        + p_pressure
        + p_credibility
    )
    probability = _clamp(raw, P_FLOOR, P_CEIL)

    breakdown: list[BreakdownItem] = [
        BreakdownItem(
            label="base rate", kind="base", delta=action.base_rate, key="base"
        ),
        BreakdownItem(
            label=_morale_label(morale_norm),
            kind="modifier",
            delta=p_morale,
            key="morale",
        ),
    ]
    if abs(p_trend) >= SIGNIFICANT_DELTA:
        breakdown.append(
            BreakdownItem(
                label=f"{region.morale_trend} trend",
                kind="modifier",
                delta=p_trend,
                key="trend",
            )
        )
    if abs(p_severity) >= SIGNIFICANT_DELTA:
        breakdown.append(
            BreakdownItem(
                label="recent event severity",
                kind="modifier",
                delta=p_severity,
                key="severity",
            )
        )
    breakdown.extend(category_items)
    if abs(p_pressure) >= SIGNIFICANT_DELTA and pressure_label:
        breakdown.append(
            BreakdownItem(
                label=pressure_label,
                kind="modifier",
                delta=p_pressure,
                key="pressure",
            )
        )
    if abs(p_credibility) >= SIGNIFICANT_DELTA and credibility_label:
        breakdown.append(
            BreakdownItem(
                label=credibility_label,
                kind="modifier",
                delta=p_credibility,
                key="credibility",
            )
        )

    explanation = _build_explanation(
        action=action,
        region=region,
        probability=probability,
        breakdown=breakdown,
    )

    return Outcome(
        action_id=action.id,
        region_id=region.region_id,
        probability=probability,
        breakdown=tuple(breakdown),
        explanation=explanation,
    )


def _pressure_delta(
    action: Action, political: PoliticalContext | None
) -> tuple[float, str | None]:
    """Aggression bias scaled by issuer pressure.

    Returns `(delta, label)`. Label is None when the contribution is below the
    significance threshold (the caller still sums delta, the row is dropped).
    """
    if political is None or political.issuer_pressure is None:
        return 0.0, None
    if action.pressure_aggression_bias == 0.0:
        return 0.0, None
    delta = action.pressure_aggression_bias * political.issuer_pressure
    if political.issuer_deadline_turns_remaining is not None:
        label = (
            f"deadline pressure (T-{political.issuer_deadline_turns_remaining})"
        )
    else:
        label = "deadline pressure"
    return delta, label


def _credibility_delta(
    action: Action, political: PoliticalContext | None
) -> tuple[float, str | None]:
    """Coercion-strength multiplier from issuer's outgoing credibility.

    Uses `immediate` (recent signal/action consistency) since coercive
    affordances depend on near-term reputation. We keep the math simple: delta
    = credibility_weight * immediate, so positive credibility helps coercive
    actions and negative credibility hurts them.
    """
    if political is None or political.bilateral_credibility_immediate is None:
        return 0.0, None
    if action.credibility_weight == 0.0:
        return 0.0, None
    delta = action.credibility_weight * political.bilateral_credibility_immediate
    src = political.issuer_faction_id or "issuer"
    dst = political.target_faction_id or "target"
    band = "low credibility" if political.bilateral_credibility_immediate < 0 else "credibility"
    label = f"{band} ({src.upper()}→{dst.upper()})"
    return delta, label


def _morale_label(morale_norm: float) -> str:
    if morale_norm >= 0.4:
        return "high morale"
    if morale_norm <= -0.4:
        return "low morale"
    return "neutral morale"


_CATEGORY_PHRASE: dict[EventCategory, tuple[str, str]] = {
    EventCategory.PROTEST: ("protest activity", "protest activity"),
    EventCategory.MILITARY_LOSS: ("recent military setbacks", "improving military posture"),
    EventCategory.ECONOMIC_STRESS: ("economic stress", "easing economic stress"),
    EventCategory.POLITICAL_INSTABILITY: ("political instability", "political stabilisation"),
    EventCategory.NATIONALIST_SENTIMENT: ("subdued nationalist sentiment", "rising nationalist sentiment"),
}


def _category_label(category: EventCategory, contribution: float) -> str:
    negative_phrase, positive_phrase = _CATEGORY_PHRASE.get(
        category, (category.value, category.value)
    )
    return negative_phrase if contribution < 0 else positive_phrase


def _build_explanation(
    *,
    action: Action,
    region: RegionIntel,
    probability: float,
    breakdown: Iterable[BreakdownItem],
) -> str:
    modifiers = [b for b in breakdown if b.kind != "base"]
    modifiers.sort(key=lambda b: abs(b.delta), reverse=True)
    top = modifiers[:2]
    pct = round(probability * 100)
    base_pct = round(action.base_rate * 100)
    if not top:
        return f"{action.name} in {region.region_id}: {pct}% (baseline)."

    direction = "Reduced" if probability < action.base_rate else "Boosted"
    if abs(probability - action.base_rate) < 0.005:
        direction = "Held at"

    phrases = [_phrase_for(item) for item in top]
    joined = " and ".join(p for p in phrases if p)
    return (
        f"{direction} from {base_pct}% baseline to {pct}% by {joined}."
        if joined
        else f"{action.name}: {pct}%."
    )


def _phrase_for(item: BreakdownItem) -> str:
    if item.delta == 0:
        return ""
    sign = "−" if item.delta < 0 else "+"
    pct = round(abs(item.delta) * 100)
    return f"{item.label} ({sign}{pct}%)"

File: axis-ref/frontend copy/src/decision/evaluator.ts
import type {
  Action,
  BreakdownItem,
  Outcome,
  PoliticalContext,
} from "@/types/decision";
import type { EventCategory, MoraleTrend, RegionIntel } from "@/types/intel";

export const SEVERITY_DIVISOR = 12.0;
export const CONTRIBUTION_DIVISOR = 8.0;
export const P_FLOOR = 0.05;
export const P_CEIL = 0.95;
export const SIGNIFICANT_DELTA = 0.015;
// Group-level scalars applied at evaluation time. Intel-side contributions
// (morale, trend, severity, category drivers) are dampened; political-side
// contributions (pressure, credibility) are amplified, so political context
// outweighs morale by roughly 3:1 in per-action max swing. The Python mirror
// in `backend/axis/decision/evaluator.py` MUST keep the same values.
export const MORALE_SCALE = 1 / 3;
export const POLITICAL_SCALE = 3.0;

const CATEGORY_PHRASE: Record<EventCategory, [negative: string, positive: string]> = {
  protest: ["protest activity", "protest activity"],
  military_loss: ["recent military setbacks", "improving military posture"],
  economic_stress: ["economic stress", "easing economic stress"],
  political_instability: ["political instability", "political stabilisation"],
  nationalist_sentiment: ["subdued nationalist sentiment", "rising nationalist sentiment"],
};

function clamp(x: number, lo: number, hi: number): number {
  return Math.max(lo, Math.min(hi, x));
}

function trendSigned(trend: MoraleTrend): number {
  if (trend === "rising") return 1;
  if (trend === "declining") return -1;
  return 0;
}

function moraleLabel(moraleNorm: number): string {
  if (moraleNorm >= 0.4) return "high morale";
  if (moraleNorm <= -0.4) return "low morale";
  return "neutral morale";
}

function categoryLabel(category: EventCategory, contribution: number): string {
  const [neg, pos] = CATEGORY_PHRASE[category] ?? [category, category];
  return contribution < 0 ? neg : pos;
}

function phraseFor(item: BreakdownItem): string {
  if (item.delta === 0) return "";
  const sign = item.delta < 0 ? "−" : "+";
  const pct = Math.round(Math.abs(item.delta) * 100);
  return `${item.label} (${sign}${pct}%)`;
}

function buildExplanation(
  action: Action,
  region: RegionIntel,
  probability: number,
  breakdown: BreakdownItem[],
): string {
  const modifiers = breakdown
    .filter((b) => b.kind !== "base")
    .sort((a, b) => Math.abs(b.delta) - Math.abs(a.delta))
    .slice(0, 2);
  const pct = Math.round(probability * 100);
  const basePct = Math.round(action.base_rate * 100);
  if (modifiers.length === 0) {
    return `${action.name} in ${region.region_id}: ${pct}% (baseline).`;
  }
  let direction: string;
  if (Math.abs(probability - action.base_rate) < 0.005) direction = "Held at";
  else direction = probability < action.base_rate ? "Reduced" : "Boosted";
  const phrases = modifiers.map(phraseFor).filter(Boolean);
  const joined = phrases.join(" and ");
  return joined
    ? `${direction} from ${basePct}% baseline to ${pct}% by ${joined}.`
    : `${action.name}: ${pct}%.`;
}

export function evaluate(
  action: Action,
  region: RegionIntel,
  political?: PoliticalContext,
): Outcome {
  const moraleNorm = (region.morale_score - 50) / 50;
  const pMorale = moraleNorm * action.morale_weight * MORALE_SCALE;

  const tSigned = trendSigned(region.morale_trend);
  const pTrend = tSigned * action.trend_weight * MORALE_SCALE;

  const severitySum = region.drivers.reduce((acc, d) => acc + d.contribution, 0);
  const severityNorm = clamp(severitySum / SEVERITY_DIVISOR, -1, 1);
  const pSeverity = severityNorm * action.severity_weight * MORALE_SCALE;

  const categoryItems: BreakdownItem[] = [];
  for (const d of region.drivers) {
    const sensitivity = action.category_sensitivities[d.category] ?? 0;
    if (sensitivity === 0) continue;
    const intensity = clamp(Math.abs(d.contribution) / CONTRIBUTION_DIVISOR, 0, 1);
    const delta = sensitivity * intensity * MORALE_SCALE;
    if (Math.abs(delta) < SIGNIFICANT_DELTA) continue;
    categoryItems.push({
      label: categoryLabel(d.category, d.contribution),
      kind: "category",
      delta,
      key: `cat:${d.category}`,
      source: "intel",
    });
  }

  const pCategories = categoryItems.reduce((a, b) => a + b.delta, 0);

  const { delta: rawPressure, label: pressureLabel } = pressureDelta(action, political);
  const { delta: rawCredibility, label: credibilityLabel } = credibilityDelta(
    action,
    political,
  );
  const pPressure = rawPressure * POLITICAL_SCALE;
  const pCredibility = rawCredibility * POLITICAL_SCALE;

  const raw =
    action.base_rate +
    pMorale +
    pTrend +
    pSeverity +
    pCategories +
    pPressure +
    pCredibility;
  const probability = clamp(raw, P_FLOOR, P_CEIL);

  const breakdown: BreakdownItem[] = [
    {
      label: "base rate",
      kind: "base",
      delta: action.base_rate,
      key: "base",
      source: "base",
    },
    {
      label: moraleLabel(moraleNorm),
      kind: "modifier",
      delta: pMorale,
      key: "morale",
      source: "intel",
      detail: `morale_norm × ${action.morale_weight.toFixed(2)} (morale weight)`,
    },
  ];
  if (Math.abs(pTrend) >= SIGNIFICANT_DELTA) {
    breakdown.push({
      label: `${region.morale_trend} trend`,
      kind: "modifier",
      delta: pTrend,
      key: "trend",
      source: "intel",
      detail: `trend × ${action.trend_weight.toFixed(2)} (trend weight)`,
    });
  }
  if (Math.abs(pSeverity) >= SIGNIFICANT_DELTA) {
    breakdown.push({
      label: "recent event severity",
      kind: "modifier",
      delta: pSeverity,
      key: "severity",
      source: "intel",
      detail: `net severity norm × ${action.severity_weight.toFixed(2)} (severity weight)`,
    });
  }
  breakdown.push(...categoryItems);
  if (Math.abs(pPressure) >= SIGNIFICANT_DELTA && pressureLabel) {
    const bias = action.pressure_aggression_bias ?? 0;
    const p = political?.issuer_pressure;
    breakdown.push({
      label: pressureLabel,
      kind: "modifier",
      delta: pPressure,
      key: "pressure",
      source: "pressure",
      detail:
        p != null
          ? `${bias.toFixed(2)} (aggression bias) × ${p.toFixed(2)} (issuer pressure)`
          : undefined,
    });
  }
  if (Math.abs(pCredibility) >= SIGNIFICANT_DELTA && credibilityLabel) {
    const w = action.credibility_weight ?? 0;
    const imm = political?.bilateral_credibility_immediate;
    breakdown.push({
      label: credibilityLabel,
      kind: "modifier",
      delta: pCredibility,
      key: "credibility",
      source: "credibility",
      detail:
        imm != null
          ? `${w.toFixed(2)} (cred. weight) × ${imm.toFixed(2)} (immediate, issuer→target)`
          : undefined,
    });
  }

  const explanation = buildExplanation(action, region, probability, breakdown);

  return {
    action_id: action.id,
    region_id: region.region_id,
    probability,
    breakdown,
    explanation,
  };
}

function pressureDelta(
  action: Action,
  political: PoliticalContext | undefined,
): { delta: number; label: string | null } {
  if (!political || political.issuer_pressure == null) return { delta: 0, label: null };
  const bias = action.pressure_aggression_bias ?? 0;
  if (bias === 0) return { delta: 0, label: null };
  const delta = bias * political.issuer_pressure;
  const label =
    political.issuer_deadline_turns_remaining != null
      ? `deadline pressure (T-${political.issuer_deadline_turns_remaining})`
      : "deadline pressure";
  return { delta, label };
}

function credibilityDelta(
  action: Action,
  political: PoliticalContext | undefined,
): { delta: number; label: string | null } {
  if (!political || political.bilateral_credibility_immediate == null) {
    return { delta: 0, label: null };
  }
  const w = action.credibility_weight ?? 0;
  if (w === 0) return { delta: 0, label: null };
  const delta = w * political.bilateral_credibility_immediate;
  const src = (political.issuer_faction_id ?? "issuer").toUpperCase();
  const dst = (political.target_faction_id ?? "target").toUpperCase();
  const band =
    political.bilateral_credibility_immediate < 0
      ? "low credibility"
      : "credibility";
  return { delta, label: `${band} (${src}→${dst})` };
}
